solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,275 @@
+
+    event WormholeTbtcReceived(address receiver, uint256 amount);
+
+    event WormholeTbtcSent(
+        uint256 amount,
+        uint16 recipientChain,
+        bytes32 recipient,
+        uint256 arbiterFee,
+        uint32 nonce
+    );
+
+    event MintingLimitUpdated(uint256 mintingLimit);
+
+    function initialize(
+        IWormholeTokenBridge _bridge,
+        IERC20Upgradeable _bridgeToken,
+        L2TBTC _tbtc
+    ) external initializer {
+        __Ownable_init();
+        __ReentrancyGuard_init();
+
+        require(
+            address(_bridge) != address(0),
+            "Wormhole bridge address must not be 0x0"
+        );
+        require(
+            address(_bridgeToken) != address(0),
+            "Bridge token address must not be 0x0"
+        );
+        require(
+            address(_tbtc) != address(0),
+            "L2TBTC token address must not be 0x0"
+        );
+
+        bridge = _bridge;
+        bridgeToken = _bridgeToken;
+        tbtc = _tbtc;
+        mintingLimit = type(uint256).max;
+    }
+
+    /// @notice This function is called when the user redeems their token on L2.
+    ///         The contract receives Wormhole tBTC representation and mints the
+    ///         canonical tBTC for the user.
+    ///         If the tBTC minting limit has been reached by this contract,
+    ///         instead of minting tBTC the receiver address receives Wormhole
+    ///         tBTC representation.
+    /// @dev Requirements:
+    ///      - The receiver of Wormhole tBTC should be the L2WormholeGateway
+    ///        contract.
+    ///      - The receiver of the canonical tBTC should be abi-encoded in the
+    ///        payload.
+    ///      - The receiver of the canonical tBTC must not be the zero address.
+    /// @param encodedVm A byte array containing a Wormhole VAA signed by the
+    ///        guardians.
+    function receiveWormhole(bytes memory encodedVm) external nonReentrant {
+        // ITokenBridge.completeTransferWithPayload completes a contract-controlled
+        // transfer of an ERC20 token. Calling this function is not enough to
+        // ensure L2WormholeGateway received Wormhole tBTC representation.
+        // Instead of going too deep into the ITokenBridge implementation,
+        // asserting who is the receiver of the token, and which token it is,
+        // we check the balance before the ITokenBridge call and the balance
+        // after ITokenBridge call. This way, we are sure this contract received
+        // Wormhole tBTC token in the given amount. This is transparent to
+        // all potential upgrades of ITokenBridge implementation and no other
+        // validations are needed.
+        uint256 balanceBefore = bridgeToken.balanceOf(address(this));
+        bytes memory encoded = bridge.completeTransferWithPayload(encodedVm);
+        uint256 balanceAfter = bridgeToken.balanceOf(address(this));
+
+        uint256 amount = balanceAfter - balanceBefore;
```

Comment:
The amount can be zero if the tokens bridged to this contract are not `bridgeToken`. So to protect against taking custody of irrelevant tokens, you may want to add `require(amount > 0, ...)`. And this would also save the user gas on a revert because amount zero is a no-op (since no new amount is actually being minted in this case).